x/dex/keeper/limit_order_book.go

```diff
@@ -0,0 +1,255 @@
+			priceTakerToMaker,
+		)
+	}
+
+	return NewLimitOrderBook0To1(
+		pair,
+		tick,
+		fillTranche,
+		placeTranche,
+	)
+}
+
+func NewLimitOrderBook1To0(
+	pair *types.TradingPair,
+	tick *types.Tick,
+	fillTranche *LimitOrderTranche,
+	placeTranche *LimitOrderTranche,
+) *LimitOrderBook {
+	return &LimitOrderBook{
+		Pair:                pair,
+		Tick:                tick,
+		LimitTrancheIndexes: tick.LimitOrderTranche1To0,
+		Tranches: map[uint64]*LimitOrderTranche{
+			tick.LimitOrderTranche1To0.FillTrancheIndex:  fillTranche,
+			tick.LimitOrderTranche1To0.PlaceTrancheIndex: placeTranche,
+		},
+	}
+}
+
+func NewLimitOrderBook0To1(
+	pair *types.TradingPair,
+	tick *types.Tick,
+	fillTranche *LimitOrderTranche,
+	placeTranche *LimitOrderTranche,
+) *LimitOrderBook {
+	return &LimitOrderBook{
+		Pair:                pair,
+		Tick:                tick,
+		LimitTrancheIndexes: tick.LimitOrderTranche0To1,
+		Tranches: map[uint64]*LimitOrderTranche{
+			tick.LimitOrderTranche0To1.FillTrancheIndex:  fillTranche,
+			tick.LimitOrderTranche0To1.PlaceTrancheIndex: placeTranche,
+		},
+	}
+}
+
+func (l *LimitOrderBook) NewTranche(trancheIndex uint64) *LimitOrderTranche {
+	token0, _ := PairToTokens(l.Pair.PairId)
+	var priceMakerToTaker, priceTakerToMaker sdk.Dec
+	if l.TokenIn == token0 {
+		priceMakerToTaker = *l.Tick.Price0To1
+	} else {
+		priceTakerToMaker = sdk.OneDec().Quo(*l.Tick.Price0To1)
+	}
+	return NewLimitOrderTranche(
+		l.Tick,
+		&types.LimitOrderTranche{
+			PairId:           l.Pair.PairId,
+			TokenIn:          l.TokenIn,
+			TickIndex:        l.Tick.TickIndex,
+			TrancheIndex:     trancheIndex,
+			ReservesTokenIn:  sdk.ZeroInt(),
+			ReservesTokenOut: sdk.ZeroInt(),
+			TotalTokenIn:     sdk.ZeroInt(),
+			TotalTokenOut:    sdk.ZeroInt(),
+		},
+		priceMakerToTaker,
+		priceTakerToMaker,
+	)
+}
+
+func (l LimitOrderBook) Swap(maxAmount sdk.Int) (
+	inAmount sdk.Int,
+	outAmount sdk.Int,
+	initedTick *types.Tick,
+	deinitedTick *types.Tick,
+) {
+	inAmount = sdk.ZeroInt()
+	outAmount = sdk.ZeroInt()
+
+	fillTranche := &l.LimitTrancheIndexes.FillTrancheIndex
+	placeTranche := &l.LimitTrancheIndexes.PlaceTrancheIndex
+
+	for maxAmount.GT(sdk.ZeroInt()) && *fillTranche < *placeTranche {
+		var curInAmount sdk.Int
+		var curOutAmount sdk.Int
+		curInAmount, curOutAmount, _, deinitedTick = l.Tranches[*fillTranche].Swap(maxAmount)
+		maxAmount = maxAmount.Sub(curInAmount)
+		inAmount = inAmount.Add(curInAmount)
+		outAmount = outAmount.Add(curOutAmount)
+		if deinitedTick != nil {
+			*fillTranche++
+		}
+	}
+
+	if maxAmount.GT(sdk.ZeroInt()) {
+		var curInAmount sdk.Int
+		var curOutAmount sdk.Int
+		curInAmount, curOutAmount, _, deinitedTick = l.Tranches[*fillTranche].Swap(maxAmount)
+		inAmount = inAmount.Add(curInAmount)
+		outAmount = outAmount.Add(curOutAmount)
+		if deinitedTick != nil {
+			*fillTranche++
+			*placeTranche++
+		} else {
+			*placeTranche++
+		}
+		l.Tranches[*placeTranche] = l.NewTranche(*placeTranche)
```

Comment:
What's the goal with this line? Doesn't seem to be necessary right now. Is there a world where we think we might be placing limit orders in the same TX that we are swapping through them?